spec/controllers/api/v0/shipments_controller_spec.rb

```diff
@@ -176,7 +176,88 @@
       end
     end
 
-    context "can transition a shipment from ready to ship" do
+    describe "#update" do
+      let!(:distributor) { create(:distributor_enterprise) }
+      let!(:shipping_method1) {
+        create(:shipping_method_with, :flat_rate, distributors: [distributor], amount: 10)
+      }
+      let!(:shipping_method2) {
+        create(:shipping_method_with, :flat_rate, distributors: [distributor], amount: 20)
+      }
+      let!(:order_cycle) { create(:order_cycle, distributors: [distributor]) }
+      let!(:order) {
+        create(:completed_order_with_totals, order_cycle: order_cycle, distributor: distributor)
+      }
+      let(:new_shipping_rate) {
+        order.shipment.shipping_rates.select{ |sr| sr.shipping_method == shipping_method2 }.first
+      }
+      let(:params) {
+        {
+          id: order.shipment.number,
+          order_id: order.number,
+          shipment: {
+            selected_shipping_rate_id: new_shipping_rate.id
+          }
+        }
+      }
+
+      before do
+        order.shipments.first.shipping_methods = [shipping_method1, shipping_method2]
+        order.select_shipping_method(shipping_method1.id)
+        order.update!
+        order.update_columns(
+          payment_total: 60,
+          payment_state: "paid"
+        )
+      end
+
+      context "when an order has multiple shipping methods available which could be chosen" do
+        context "changing the selected shipping method" do
+          it "updates the order's totals and states" do
+            expect(order.shipment.shipping_method).to eq shipping_method1
+            expect(order.shipment.cost).to eq 10
+            expect(order.total).to eq 60 # item total is 50, shipping cost is 10
+            expect(order.payment_state).to eq "paid" # order is fully paid for
```

Comment:
I don't understand why we need to make assertions on the test setup. If this fails it's because either the factory or the test setup is broken and we should see that when implementing the test. I personally never liked this approach. It feels like unconfident programming that raises the cost of changes to these tests in the future.

Reply:
I've hit issues multiple times whilst working on adjustments where code changes meant that the factories themselves were not working as intended, and specs were failing due to the test setup not being completed correctly as opposed to the actual code under test being broken. In those cases it took a really long time to debug.

I see what you mean, but in this case where we need to really carefully check a number of attributes both before and after (and the changes are not trivial) I feel like being really explicit and clear here about exactly what's changing is actually a good thing.

The test setup here is expensive, but checking all these attributes is really cheap, right?

Reply:
Interesting discussion! I've ran into the scenario (can't remember where) in which I "felt better" following this approach as well. Mainly because I'm starting to use factories and wanted to make sure I had the correct setup to kick off with testing the relevant features, with following assertions.

The takeaway seems to be: it's not something we should adopt as standard, but it can be useful (and cheap/fast) to make sure that we've nailed the starting point, before we actually run the assertions we wish to test.